File: app/src/main/java/com/silentgames/silent_planet/model/cells/CellType.java
package com.silentgames.silent_planet.model.cells;

import com.silentgames.silent_planet.model.CellEx;
import com.silentgames.silent_planet.model.cells.defaultCell.Default;
import com.silentgames.silent_planet.model.cells.onVisible.OnVisible;

/**
 * Created by gidroshvandel on 09.07.16.
 */
public class CellType extends CellTypeEx  {

    private Default aDefault;
    private OnVisible onVisible;

    public CellType(OnVisible onVisible) {
        this.onVisible = onVisible;
        setAll(onVisible);
    }

    public CellType(Default aDefault) {
        this.aDefault = aDefault;
        setAll(aDefault);
    }

    public OnVisible getOnVisible() {
        return onVisible;
    }

    public void setOnVisible(OnVisible onVisible) {
        this.onVisible = onVisible;
    }

    public Default getDefault() {
        return aDefault;
    }

    public void setDefault(Default aDefault) {
        this.aDefault = aDefault;
    }

}

File: app/src/main/java/com/silentgames/silent_planet/model/cells/CellType.kt
package com.silentgames.silent_planet.model.cells

import android.content.Context
import android.graphics.Bitmap
import com.silentgames.silent_planet.R
import com.silentgames.silent_planet.model.Event
import com.silentgames.silent_planet.utils.BitmapEditor

/**
 * Created by gidroshvandel on 09.07.16.
 */
abstract class CellType(
        val context: Context,
        override var crystals: Int = 0,
        override val closeBitmap: Bitmap = BitmapEditor.getCellBitmap(context, R.drawable.planet_background),
        override var isCanFly: Boolean = false,
        override var isDead: Boolean = false,
        override var isVisible: Boolean = false,
        name: String = "",
        description: String = ""
) : CellTypeProperties {

    override var name: String = name
        get() = if (isVisible) field else context.getString(R.string.unknown_cell_name)

    override var description: String = description
        get() = if (isVisible) field else context.getString(R.string.unknown_cell_description)

    open fun doEvent(event: Event) {
    }

    fun getCurrentBitmap(): Bitmap = if (isVisible) {
        bitmap
    } else {
        closeBitmap
    }
}
